llvm/include/llvm/TableGen/ScanDependencies.h

```diff
@@ -78,11 +79,11 @@ struct ScanIncludesResult {
 /// TODO: We should upstream a version of this that doesn't rely on the CAS;
 /// it'd be useful for build systems that want to collect dependencies ahead of
 /// time. Maybe can separate a tool/etc. for that.
-Expected<ScanIncludesResult>
-scanIncludes(cas::CASDB &CAS, cas::CASID ExecID, StringRef MainFilename,
-             ArrayRef<std::string> IncludeDirs,
-             ArrayRef<MappedPrefix> PrefixMappings = None,
-             Optional<TreePathPrefixMapper> *CapturedPM = nullptr);
+Expected<ScanIncludesResult> scanIncludes(
+    cas::CASDB &CAS, cas::CASID ExecID, StringRef MainFilename,
+    ArrayRef<std::string> IncludeDirs, cas::CachingOnDiskFileSystem &FS,
+    function_ref<StringRef(const vfs::CachedDirectoryEntry &)> RemapPath =
```

Comment:
Experiment a bit here:
* Using `llvm::is_trivially_move_constructible` fixes the problem
* I can't seem to produce a unit test for this as `Optional<BumpPtrAllocator>` works fine and few other simple example I tried. Probably need to have a good and deep understanding of the compiler "bug" to write a test case if I want to fix this upstream. @dexonsmith Any suggestions?

Reply:
Great news, glad that works.

For the test, probably `Optional<BumpPtrAllocator>` on its own reduced the testcase too much.
```
struct ImplicitlyNonMovable {
  ImplicitlyNonMovable(); // Constructor out-of-line.
  Optional<BumpPtrAllocator> Alloc;
};
void func() { Optional<ImplicitlyNonMovable> V; }
```
If that doesn't trigger it either, maybe try adding parameters to the constructor, or adding a third layer (since the original code had three):
```
struct T1 {
  T1();
  Optional<BumpPtrAllocator> Alloc;
};
struct T2 {
  T2();
  T1 M1;
};
void func() { Optional<T2> V; }
```
Also, probably better to use a minimal type rather than BumpPtrAllocator in the test for upstream, but you can reduce that later.

BTW, for the purposes of this PR (getting the bots going), I don't think you need a test. Just fix Optional with a FIXME to add a test. (But I know you need one for the upstream patch.)

Reply:
Oh, this comment might point at a failing example: https://reviews.llvm.org/D93510#2538983

Looks at least like you could copy the godbolt example and iterate there, if that's not already what you're doing.

Reply:
Unfortunately I tried all of those. If I have time later, I might try to creduce a test case.

Reply:
Did you try the `virtual ~... = default` suggested in the comment I just linked to?

Reply:
Here's a link to a repro https://godbolt.org/z/c4oso1TTe

Reply:
This is a more accurate repro https://godbolt.org/z/xbP68YG47

Reply:
Ah, that is the piece I am missing. Here you go: https://reviews.llvm.org/D117254

Reply:
Manually reduced on godbolt (https://godbolt.org/z/934cbao4T):
```
struct NoCopy {
  NoCopy() = default;
  NoCopy(const NoCopy &) = delete;
};
struct HasNoCopyOptional {
  ~HasNoCopyOptional();
  Optional<NoCopy> Alloc;
};

void func() { Optional<HasNoCopyOptional> O; }
```

Reply:
Confirmed:
- Fails in GCC 5.4, unless you switch to `llvm::is_trivially_move_constructible`
- Works in clang
- Works in newer GCC